ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,348 @@
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Summary
+
+Proposal to improve developer-experience by allowing to upload a local repository clone to the
+Kubernetes cluster, and use Shipwright Build Controller to produce a container-image out of it.
+
+As a developer, I can try Shipwright before opening a pull-request, and take advantage of the cluster
+resources to create the image. Effectively moving shipwright closer to the developer's inner-loop, in
+other words closer to the frequent activities while writing code ([1][InnerLoopA], [2][InnerLoopB]),
+in other words we are "moving left before _shipping-right_".
+
+## Motivation
+
+Shipwright Build-Controller allows teams to build container-images out of Git repositories, register
+and track their application builds in a Kubernetes native environment. This enhancement proposal
+extends this ability for developers to upload uncommitted changes and use the same machinery during
+the inner-loop as well.
+
+As a developer you can try out changes before opening a pull-request, make sure the final
+container-image will work out as expected, and effectively improve the developer experience.
+
+The approach proposed here is complementary to existing [EP][LocalSourceCodeEP] detailing local
+source upload to a container-registry and later on self-extracting the image before the actual
+container build process starts.
+
+### Goals
+
+- Improve developer experience, local clone can produce a container-image using Shipwright machinery
+- Move Shipwright CLI closer to developer's inner-loop
+- Complementary to [upload-to-container-registry (EP)][LocalSourceCodeEP], uploading data directly to
+the cluster instead
+
+## Proposal
+
+This proposal includes changes in Shipwright components, and adds a new helper application to aid
+data upload process.
+
+The user-experience using the CLI should be as simple as the following example, assuming the user is
+present at the same directory than the source code, and a `Build` named `nodejs-ex` exists.
+
+```bash
+shp buildrun upload nodejs-ex --to-cluster
+```
+
+Here, a new sub-command, `upload` is introduced and the user may choose `--to-cluster` or
+`--to-registry` options, to respectively, upload to the cluster directly or upload to
+container-registry.
+
+### Sequence Diagram
+
+The following sequence-diagram shows the relationship between the components and explains the order
+of execution. In a nutshell, the CLI creates a `BuildRun` object, with settings to wait for data
+upload, which can only start when the Build-Controller instructs Kubernetes to create this new pod.
+The new pod, on the other hand, needs to be running and ready to receive external data.
+
+With the data upload complete, the build-pod can continue to execute the build-strategy chosen,
+producing a container-image at the end.
+
+![source-upload sequence diagram](assets/0005-source-upload-sequence-diagram.png)
```

Comment:
This image has a transparent background. I am using GitHub in dark mode. When reading further I wondered about the step numbers that you reference until I eventually noticed that they are defined on the image. You may want to update the image with a white background.

Reply:
Sure! I've just pushed a new image with white background. Thanks.